[dagster-airflow] [docs] migration guide updates/considerations (#12198)
dagster-io/dagster: docs/content/integrations/airflow/migrating-to-dagster.mdx

You'll also need to configure the [Airflow connections](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html) that your DAGs depend on. To accomplish this, use the `connections` parameter instead of URI-encoded environment variables.

```python file=/integrations/airflow/migrate_repo_connections.py
import os

from airflow.models import Connection
from dagster_airflow import make_dagster_definitions_from_airflow_dags_path

migrated_airflow_definitions = make_dagster_definitions_from_airflow_dags_path(
    os.path.abspath("./dags/"),
    connections=[Connection(conn_id="http_default", conn_type="uri", host="https://google.com")],
)
```

Iterate as needed until all configuration is correctly ported to your local environment.

---

## Step 6: Move to production

<Note>
  This step is applicable to Dagster Cloud. If deploying to your infrastructure,
  refer to the <a href="/deployment">Deployment guides</a> for more info.
  <br /><br />
  Additionally, until your Airflow DAGs execute successfully in your local environment, we recommend waiting to move to production.
</Note>

In this step, you'll set up your project for use with Dagster Cloud.

> <!-- Once you have your dagster cloud organization setup you can either choose to adapt your existing airflow repository to use the Dagster Cloud CI/CD deployment or copy over your code to the quickstart repository you made. -->

Reviewer: ❓ Pulled this from some commits made after I started.

I'm not sure exactly what this means, specifically the `quickstart repository you made`. What quickstart repo? The one from the Dagster Cloud GS guide?

Author: yes the quickstart repo that you made in the cloud GS guide

   ```

---

## Step 7: Migrate permissions to Dagster

Your Airflow instance likely had specific IAM or Kubernetes permissions that allowed it to successfully run your Airflow DAGs. To run the migrated Dagster jobs, you'll need to duplicate these permissions for Dagster.

- **We recommend using [Airflow connections](https://airflow.apache.org/docs/apache-airflow/stable/howto/connection.html) or [environment variables](/dagster-cloud/developing-testing/environment-variables-and-secrets)** to define permissions whenever possible.

- **If you're unable to use Airflow connections or environment variables,** you can attach permissions directly to the infrastructure where you're deploying Dagster.

- **If your Airflow DAGs used [`KubernetesPodOperators`](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/operators.html)**, it's possible that you loaded a `kube_config` file or used the `in_cluster` config. When migrating, we recommend switching to [using connections with a `kube_config` JSON blob](https://airflow.apache.org/docs/apache-airflow-providers-cncf-kubernetes/stable/connections/kubernetes.html) to make things easier.

---

> ## Containerized Operator Considerations

Reviewer: ❓ Is this relevant to a specific step in this guide, or just migration in general? If it's specific to a step (perhaps the local dev portion of the guide?), consider moving it there. A user might not see it otherwise, or see it after the fact and have a bad time.

LMK your thoughts - I didn't edit this because I wanted info first, but I'd like to take a pass at it before this gets merged.

Author: yeah let me move this up into the running dagit sectioin

This guide will walk you through the steps of performing this migration.

---

## Prerequisites

Migrations are complicated as there are lots of moving pieces. It's important for you to go in knowing exactly how much scope and complexity you'll be tackling.

To complete the migration, you'll need:

- **To know the following about your Airflow setup**:

  - What operator types are used in the DAGs you're migrating
  - What Airflow connections your DAGs depend on
  - What Airflow variables you've set
  - What Airflow secrets backend you use
  - Where the permissions that your DAGs depend on are defined

  If you're unsure where to find these things, you're likely not the right person to make the migration. We recommend looping a teammate in before continuing.

- **If using Dagster Cloud**, an existing [Dagster Cloud](/dagster-cloud) account. While your migrated Airflow DAGs will work with Dagster Open Source, this guide includes setup specific to Dagster Cloud.

  **If you just signed up for a Cloud account**, follow the steps in the [Dagster Cloud Getting Started guide](/dagster-cloud/getting-started) before proceeding.

---

> ## Limitations

Reviewer: I moved this up because I thought users would probably want to know about any limitations before starting the migration. Thoughts?

Author: 👍 makes sense

```python file=/integrations/airflow/migrate_repo.py
import os

from dagster_airflow import (
    make_dagster_definitions_from_airflow_dags_path,
)

migrated_airflow_definitions = make_dagster_definitions_from_airflow_dags_path(
    os.path.abspath("./dags/"),
)
```

---

## Step 4: Verify the DAGs are loading

In this step, you'll spin up Dagit, Dagster's web-based UI, and verify that your migrated DAGs are loading. **Note**: Unless the migrated DAGs are very, very simple, it's unlikely they'll execute correctly at this point. That's okay - we'll fix it in a bit. Starting Dagit is the first step in our development loop, allowing you to make a local change, view it in Dagit, and debug any errors.

1. Run the following to start Dagit:

   ```bash
   dagster dev -f ./migrate_repo.py
   ```

2. In your browser, navigate to <http://localhost:3001>. You should see a list of Dagster jobs that correspond to the DAGs in your Airflow DagBag.

3. Run one of the simpler jobs, ideally one where you're familiar with the business logic. Note that it's likely to fail due to a configuration or permissions issue.

> 4. Using `stdout/stderr`, identify and address the cause of the failure.

Reviewer: ❓ Does this refer to the logs? I kept it from the original draft, but it felt odd to me. If it's the logs, we should refer users to them in Dagit and mention that they can view `stdout/stderr` there if they want.

Author: I'll change it to logs, but specifically these are the compute logs not our dagster event log. I've seen the difference be a point of confusion for users

Reviewer: Understood! I think just telling users where to look is the important thing here. I assumed 'logs' in general because the instructions had me in Dagit at the time.

  !
</Note>

Dagster can convert your Airflow DAGs into Dagster jobs, enabling a lift-and-shift migration from Airflow without any rewriting.

This guide will walk you through the steps of performing this migration.

---

## Prerequisites

> Before beginning a migration it is important for you to do some upfront analysis:
> 
> - **To understand the complexity of undertaking a migration**. Refer to the [previous section](#before-you-begin) for more detail.
> 

Reviewer: This reads strangely now, because the first sentence in this section doesn't relate/lead into every one of the bullet points in this section.

Also changing `previous section` to `next section` since that's where it wound up being placed.

Suggestion:

```suggestion
To complete the migration, you'll need:

- **To perform some upfront analysis**. Refer to the [next section](#before-you-begin) for more detail.

```

- **To know the following about your Airflow setup**:

  - What operator types are used in the DAGs you're migrating
  - What Airflow connections your DAGs depend on
  - What Airflow variables you've set
  - What Airflow secrets backend you use
  - Where the permissions that your DAGs depend on are defined

- **If using Dagster Cloud**, an existing [Dagster Cloud](/dagster-cloud) account. While your migrated Airflow DAGs will work with Dagster Open Source, this guide includes setup specific to Dagster Cloud.

  **If you just signed up for a Cloud account**, follow the steps in the [Dagster Cloud Getting Started guide](/dagster-cloud/getting-started) before proceeding.

> ---

Reviewer: Could you please remove this? Rules are only needed before H2s.

)

migrated_airflow_definitions = make_dagster_definitions_from_airflow_dags_path(
    os.path.abspath("./dags/"),
)
```

---

## Step 4: Verify the DAGs are loading

In this step, you'll spin up Dagit, Dagster's web-based UI, and verify that your migrated DAGs are loading. **Note**: Unless the migrated DAGs depend on no Airflow configuration state or permissions, it's unlikely they'll execute correctly at this point. That's okay - we'll fix it in a bit. Starting Dagit is the first step in our development loop, allowing you to make a local change, view it in Dagit, and debug any errors.

1. Run the following to start Dagit:

   ```bash
   dagster dev -f ./migrate_repo.py
   ```

2. In your browser, navigate to <http://localhost:3001>. You should see a list of Dagster jobs that correspond to the DAGs in your Airflow DagBag.

3. Run one of the simpler jobs, ideally one where you're familiar with the business logic. Note that it's likely to fail due to a configuration or permissions issue.

4. Using logs to identify and making configuration changes to fix the cause of the failure.

Repeat these steps as needed until the jobs run successfully.

> #### Containerized Operator Considerations

Reviewer: ```suggestion
### Containerized operator considerations
```

This should be an H3, and headings use sentence casing

  !
</Note>

Dagster can convert your Airflow DAGs into Dagster jobs, enabling a lift-and-shift migration from Airflow without any rewriting.

This guide will walk you through the steps of performing this migration.

---

## Prerequisites

> Migrations are complicated as there are lots of moving pieces. It's important for you to go in knowing exactly how much scope and complexity you'll be tackling.

Reviewer 2: I don't think we need to scare people right out of gate.

Suggestion:

"Before beginning a migration it is important for you to do some upfront analysis:"

Then drop, "To complete the migration, you'll need:"

To complete the migration, you'll need:

- **To understand the complexity of undertaking a migration**. Refer to the [previous section](#before-you-begin) for more detail.

- **To know the following about your Airflow setup**:

  - What operator types are used in the DAGs you're migrating
  - What Airflow connections your DAGs depend on
  - What Airflow variables you've set
  - What Airflow secrets backend you use
  - Where the permissions that your DAGs depend on are defined

>   If you're unsure where to find these things, you're likely not the right person to make the migration. We recommend looping a teammate in before continuing.

Reviewer 2: I think we can drop "If you're unsure where to find these things, you're likely not the right person to make the migration. We recommend looping a teammate in before continuing."

For larger installations, teams that already adopt devops practices and/or have standardized usage of Airflow will have a smoother transition.

Some concrete examples:

- If you rely on the usage of the Airflow UI to set production connections or variables, this will require a change of workflow as you will no longer have the Airflow UI at the end of this migration. If instead you rely on code-as-infrastructure patterns to set connections or variables, migration is more straightforward.
- If you have standardized on a small set or even a single operator type (e.g. the `K8sPodOperator`), migration will be easier. If you use a large number of operator types with a wide range of infrastructure requirements, migration will be more work.
- If you dynamically generate your Airflow DAGs from a higher-level API or DSL (e.g. yaml), the migration will be more straightforward than all your stakeholders directly creating Airflow DAGs.

> Even in the case that requires some infrastructure investment, `dagster-airflow` dramatically eases migration, typically by orders of magnitude. The cost can be borne by a single or small set of infrastructure-oriented engineers, as it dramatically reduces coordination costs: You do not have to move all of your stakeholders over to new APIs simultaneously. In our experience, practitioners welcome the change, because of the immediate improvement in tooling, stability, and development speed.

Reviewer 2: "The cost can be borne by a single or small set of infrastructure-oriented engineers, which dramatically reduces coordination costs"

---

## Limitations

Before kicking off a migration, note that there are some limitations to converting Airflow DAGs into Dagster jobs, assets, and schedules. The following features are currently unsupported:

- `retry-from-failure` in Dagster
- Using `prev_execution_date` in Airflow-templated DAGs
- `SubDAGOperators`
- Airflow Datasets
- Airflow Pools

> Most of the current limitations are due to a locally-scoped, ephemeral Airflow database being created for each run. If your DAGs implicitly rely on Airflow database state across `DagRuns`, then the current migration tooling won't work for you.

Reviewer 2: Let's link to an issue here which itself backlinks to this so we make sure to eliminate this text ASAP.

To install everything, run:

```bash
pip install dagster dagster-airflow dagit
```

We also suggest verifying that you're installing the correct versions of your Airflow dependencies. Verifying the dependency versions will likely save you from debugging tricky errors later.

To check dependency versions, open your Airflow provider's UI and locate the version numbers. When finished, continue to the next step.

---

## Step 3: Convert DAGS into Dagster definitions

In this step, you'll start writing Python!

In the `dagster_migration.py` file you created in [Step 1](#step-1-prepare-your-project-for-a-new-dagster-python-module), use <PyObject module="dagster_airflow" object="make_dagster_definitions_from_airflow_dags_path" /> and pass in the file path of your Airflow DagBag. Dagster will load the DagBag and convert all DAGs into Dagster jobs and schedules.

```python file=/integrations/airflow/migrate_repo.py
import os

from dagster_airflow import (
    make_dagster_definitions_from_airflow_dags_path,
)

migrated_airflow_definitions = make_dagster_definitions_from_airflow_dags_path(
    os.path.abspath("./dags/"),
)
```

---

## Step 4: Verify the DAGs are loading

> In this step, you'll spin up Dagit, Dagster's web-based UI, and verify that your migrated DAGs are loading. **Note**: Unless the migrated DAGs are very, very simple, it's unlikely they'll execute correctly at this point. That's okay - we'll fix it in a bit. Starting Dagit is the first step in our development loop, allowing you to make a local change, view it in Dagit, and debug any errors.

Reviewer 2: If it is easy to be more specific than "very, very simple" that would be good.

Reviewer: Agreed - I did this late last night and am almost never a fan of qualifiers.

We could drop them and just go with `simple`, or elaborate and define what *we* mean by simple.

1. Run the following to start Dagit:

   ```bash
   dagster dev -f ./migrate_repo.py
   ```

2. In your browser, navigate to <http://localhost:3001>. You should see a list of Dagster jobs that correspond to the DAGs in your Airflow DagBag.

3. Run one of the simpler jobs, ideally one where you're familiar with the business logic. Note that it's likely to fail due to a configuration or permissions issue.

4. Using logs to identify and making configuration changes to fix the cause of the failure.

Repeat these steps as needed until the jobs run successfully.

#### Containerized Operator Considerations

> There are a variety of Airflow Operator types that are used to create container runs in various container execution environments, for example kubernetes or amazon ECS, when getting things working locally we'd recommend trying to execute those containers locally whenever possible and only relying on remote execution when its either unrealistic or impossible to locally emulate the cloud environment. For doing local execution it likely means that you will need to have a local kubernets cluster running, we'd recommnd docker's built in k8s environment, and to be able to pull down the container images that will be needed for execution to your local machine.

Reviewer 2: I did a pass for typos and added a last sentence about Cloud:

There are a variety of Airflow Operator types that are used to launch compute in various external execution environments, for example Kubernetes or Amazon ECS. When getting things working locally we'd recommend trying to execute those containers locally unless it's either unrealistic or impossible to emulate the cloud environment. For example is you use the K8sPodOperator, it likely means that you will need to have a local Kubernetes cluster running, and in that case we recommend docker's built-in Kubernetes environment. You also need to be able to pull down the container images that will be needed for execution to your local machine.

If local execution is impossible, we recommend using Branch Deployments in Dagster Cloud, which is a well-supported workflow for cloud-native development.

Reviewer 2: ^---- Re: this are we sure we want to recommend this? I highly doubt that most people's containers can work on a laptop, even if they get a local k8s cluster going. We might be leading people down the wrong path.

---

## Step 1: Take inventory of your existing Airflow usage

> Making migrations is complicated, there are lots of moving pieces and it will be important for you to go in knowing exactly how much scope and complexity you'll be tackling.

Reviewer 2: Maybe something like:

"Before embarking on a migration it is important to understand the scope and complexity of your existing Airflow installation."

> There are a few things you'll want to do ahead of time, specifically you'll want to make a note of:

Reviewer 2: @Ramshackle-Jamathon I was thinking something in this tone:

### How easy is this going to be, really?

You may be coming to this document/library in a skeptical frame of mind, having previously been burned by projects that claim to have 100% foolproof, automated migration tools. We want to assure you that we are _not_ making that claim.

The `dagster-airflow` migration library should be viewed as a powerful _accelerant_ of migration, rather than guaranteeing completely _automated_ migration. The amount of work required is proportional to the complexity of your installation. Smaller implementations of less complexity can be trivial to migrate, making it appear virtually automatic; larger, more complicated, or customized implementations require more investment. 

For larger installations, teams that already adopt devops practices and/or have standardized usage of Airflow will have a smoother transition. 

Some concrete examples:

* If you rely on the usage of the Airflow UI to set production connections or variables, this will require a change of workflow as you will no longer have the Airflow UI at the end of this migration. If instead you rely on code-as-infrastructure patterns to set connections or variables, migration more straightforward. 
* If you have standardized on a small set or even a single operator type (e.g. the K8sPodOperator), migration will be easiser. If you use a large number of operator types with a wide range of infrastructure requirements, migration will be more work.
* If you dynamically generate your Airflow DAGs from a higher-level API or DSL (e.g. yaml), the migration will be more straightforward than all your stakeholders directly creating Airflow DAGs.

Even in the case that requires some infrastructure investment, `dagster-airflow` dramatically eases migration, by orders of magnitude. The cost can be borne by a single or small set of infrastructure-oriented engineers, and it dramatically reduces coordination costs, as you do not have to move all of your stakeholders over to new APIs simultaneously. In our experience, practitioners welcome the change, because of the immediate improvement in tooling, stability, and development speed.

Author: ah, do you mind if I lift this directly into the docs?
